File: src/main/host/syscall/random.c
/*
 * The Shadow Simulator
 * See LICENSE for licensing information
 */

#include "main/host/syscall/random.h"

#include <errno.h>

#include "main/host/host.h"
#include "main/host/thread.h"
#include "main/host/syscall/protected.h"
#include "main/utility/random.h"
#include "support/logger/logger.h"


///////////////////////////////////////////////////////////
// System Calls
///////////////////////////////////////////////////////////

SysCallReturn syscallhandler_getrandom(SysCallHandler* sys,
                                   const SysCallArgs* args) {
    PluginPtr bufPtr = args->args[0].as_ptr; // char*
    size_t count = args->args[1].as_u64;
    // We ignore the flags arg, because we use the same random source for both
    // random and urandom, and it never blocks anyway.

    debug("Trying to read %zu random bytes.", count);

    if(!bufPtr.val) {
        info("Invalid buffer.");
        return (SysCallReturn){.state = SYSCALL_DONE, .retval.as_i64 = -EFAULT};
    }

    // Get the buffer where we can copy the random bytes
    char* buf = memorymanager_getWriteablePtr(sys->memoryManager, sys->thread, bufPtr, count);

    // Get the source from the host to maintain determinism.
    Random* rng = host_getRandom(sys->host);
    utility_assert(rng != NULL);
    
    // We always get the number of bytes we requested.
    random_nextNBytes(rng, buf, count);
    return (SysCallReturn){.state = SYSCALL_DONE, .retval.as_i64 = count};
}

File: src/main/host/syscall/handler/random.rs
use linux_api::errno::Errno;
use log::*;
use rand::Rng;
use shadow_shim_helper_rs::syscall_types::ForeignPtr;

use crate::host::syscall::handler::{SyscallContext, SyscallHandler};
use crate::host::syscall::types::ForeignArrayPtr;

impl SyscallHandler {
    log_syscall!(
        getrandom,
        /* rv */ isize,
        /* buf */ *const std::ffi::c_void,
        /* count */ usize,
        /* flags */ std::ffi::c_uint,
    );
    pub fn getrandom(
        ctx: &mut SyscallContext,
        buf_ptr: ForeignPtr<u8>,
        count: usize,
        _flags: std::ffi::c_uint,
    ) -> Result<isize, Errno> {
        // We ignore the flags arg, because we use the same random source for both
        // random and urandom, and it never blocks anyway.

        trace!("Trying to read {count} random bytes.");

        // Get random bytes using host rng to maintain determinism.
        let rnd_bytes = {
            let mut v = vec![0u8; count];
            let mut rng = ctx.objs.host.random_mut();
            rng.fill_bytes(&mut v);
            v
        };

        // Copy to managed process.
        ctx.objs
            .process
            .memory_borrow_mut()
            .copy_to_ptr(ForeignArrayPtr::new(buf_ptr, count), &rnd_bytes)?;

        Ok(isize::try_from(count).unwrap())
    }
}
